saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py

```diff
@@ -0,0 +1,322 @@
+                    {
+                        "instance": instance,
+                        "errors": index_error_map[index],
+                        "cleaned_input": cleaned_input,
+                    }
+                )
+            except ValidationError as exc:
+                for key, value in exc.error_dict.items():
+                    index_error_map[index].append(
+                        ProductVariantBulkError(
+                            field=to_camel_case(key),
+                            message=exc.message,
+                            code=exc.code,
+                        )
+                    )
+                instances_data_and_errors_list.append(
+                    {"instance": None, "errors": index_error_map[index]}
+                )
+        return instances_data_and_errors_list
+
+    @classmethod
+    @traced_atomic_transaction()
+    def update_or_create_variant_stocks(cls, variant, cleaned_input, manager):
+        stocks = []
+        if stocks_data := cleaned_input.get("stocks"):
+            for stock_data in stocks_data:
+                stock, is_created = warehouse_models.Stock.objects.get_or_create(
+                    product_variant=variant, warehouse=stock_data["warehouse"]
+                )
+
+                if is_created or (stock.quantity <= 0 and stock_data["quantity"] > 0):
+                    transaction.on_commit(
+                        lambda: manager.product_variant_back_in_stock(stock)
+                    )
+
+                if stock_data["quantity"] <= 0:
+                    transaction.on_commit(
+                        lambda: manager.product_variant_out_of_stock(stock)
+                    )
+
+                stock.quantity = stock_data["quantity"]
+                stocks.append(stock)
+
+            warehouse_models.Stock.objects.bulk_update(stocks, ["quantity"])
+
+    @classmethod
+    def update_or_create_variant_channel_listings(cls, variant, cleaned_input):
+        channel_listings_data = cleaned_input.get("channel_listings")
+        if not channel_listings_data:
+            return
+
+        for channel_listing_data in channel_listings_data:
+            channel = channel_listing_data["channel"]
+            price = channel_listing_data["price"]
+            cost_price = channel_listing_data.get("cost_price")
+            preorder_quantity_threshold = channel_listing_data.get("preorder_threshold")
+
+            defaults = {
+                "price_amount": price,
+                "currency": channel.currency_code,
+            }
+
+            if "preorder_threshold" in channel_listing_data:
+                defaults["preorder_quantity_threshold"] = preorder_quantity_threshold
+
+            if "cost_price" in channel_listing_data:
+                defaults["cost_price_amount"] = cost_price
+
+            models.ProductVariantChannelListing.objects.update_or_create(
+                variant=variant, channel=channel, defaults=defaults
+            )
```

Comment:
I do not like this 😄 . We should take care of creation and update separately to do this in bulk